### ISSUE TITLE
Garbage collection performance degrades as database file size increases

### ISSUE THREAD

**Issue by bostrt** [SpiderLabs/ModSecurity#1286]
As the database file size increases, garbage collection duration increases and causes request slowness. Here is a configuration that reproduces the issue and a script to generate requests to bloat the database size. At the end of our 10,000 request test, the database size was only 49M.

Make sure you do not have `SecCollectionTimeout` configured so the issue can be reproduced quickly.

````
# access_log format
LogFormat "%t \"%r\" %>s %b %{SESSIONID}C Time Taken: %D" combined

<IfModule mod_security2.c>
    SecRuleEngine On
    SecDebugLog logs/modsec_debug.log
    SecDebugLogLevel 9
    SecCookieFormat 0
    SecTmpDir /dev/shm/mod_security/
    SecDataDir /dev/shm/mod_security/

    SecRule REQUEST_COOKIES:SESSIONID !^$  phase:1,id:118,nolog,pass,setsid:%{REQUEST_COOKIES.SESSIONID}
    SecAction phase:1,id:119,nolog,pass,setvar:SESSION.my_counter=+1
</IfModule>
````

Here is the "client" script:
````
#!/bin/bash
for i in `seq 1 10000`;do
    curl -b "SESSIONID=test$i" localhost/index.html &>/dev/null
done
````

To reproduce, start httpd with the configuration above, run the client script, and then observe the access_log and mod_security debug logs to see the issue happening. 

Here is a snippet from access_log showing when the slowness starts to happen:

````
[14/Dec/2016:08:38:03 -0500] "GET /index.html HTTP/1.1" 200 test2736 Time Taken: 1512
[14/Dec/2016:08:38:03 -0500] "GET /index.html HTTP/1.1" 200 test2737 Time Taken: 67055
[14/Dec/2016:08:38:03 -0500] "GET /index.html HTTP/1.1" 200 test2738 Time Taken: 121765
[....]
[14/Dec/2016:08:38:53 -0500] "GET /index.html HTTP/1.1" 200 6 test8480 Time Taken: 1789
[14/Dec/2016:08:38:53 -0500] "GET /index.html HTTP/1.1" 200 6 test8481 Time Taken: 213163
[14/Dec/2016:08:38:53 -0500] "GET /index.html HTTP/1.1" 200 6 test8482 Time Taken: 435829
[14/Dec/2016:08:38:53 -0500] "GET /index.html HTTP/1.1" 200 6 test8483 Time Taken: 428476
````

Finally, here is an example of the mod_security debug logging showing garbage collection duration for a long-running request:

````
[14/Dec/2016:08:38:53 --0500] [localhost/sid#7f0be55d13b0][rid#7f0be569fce0][/index.html][4] Garbage collection took 228151 microseconds.
````

**Comment by zimmerle** [comment]
Hi @bostrt 

What about your CPU during the tests? What about your I/O performance? The huge collection may lead to a CPU usage increase. Same for I/O.

Usually the cleanup happens during the logging phase, it means that the user already received the response by the time that we process the cleanup. However, the "worker" is still busy doing the cleanup and not able to answer to a new request. Are you using process or threads? how many?

**Comment by bostrt** [comment]
It seems that frequent scans of a large database are causing the issue I am seeing. Sometimes scans are unnecessary and still have to check entire set of collections so I'm currently checking on performing less frequent scans or making it a configurable option. 

Also, looks like moving from `rand` to `ap_random_pick` improved the situation (see link below) when 2+ workers check if GC should run at the same time. From what I understand, `rand` is not thread safe and `ap_random_pick` could be used for added safety. I'll work on a patch for this after some more confirmation. 

https://github.com/SpiderLabs/ModSecurity/blob/master/apache2/modsecurity.c#L240

I was typing this out right when you updated :) I did monitor hardware while running tests and CPU usage increased significantly while GC scan was being performed (expected).

> However, the "worker" is still busy doing the cleanup and not able to answer to a new request

Yes, that is part of the issue. As the database file gets larger, more and more workers are hung up on the scan.

**Comment by zimmerle** [comment]
Closed as part of #1289. Indeed, scan the database seems to be a jobs for another process (out of band). Or the expiration should be checked due to the utilization; That may be something for v3 :)
